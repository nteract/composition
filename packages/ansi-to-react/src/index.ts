import * as React from "react";
import { ansiToJson, AnserJsonEntry } from "anser";
import { escapeCarriageReturn } from "escape-carriage";

const LINK_REGEX = /(https?:\/\/(?:www\.|(?!www))[^\s\.]+\.[^\s]{2,}|www\.[^\s]+\.[^\s]{2,})/;

/**
 * Converts ANSI strings into JSON output.
 * @name ansiToJSON
 * @function
 * @param {String} input The input string.
 * @return {Array} The parsed input.
 */
function ansiToJSON(input: string) {
  input = escapeCarriageReturn(input);
  return ansiToJson(input, {
    json: true,
    remove_empty: true
  });
}

/**
 * Converts an Anser bundle into a React Node.
 * @param linkify whether links should be converting into clickable anchor tags.
 * @param bundle Anser output.
 */
function convertBundleIntoReact(
  linkify: boolean,
  bundle: AnserJsonEntry,
  key: number
) {
  const style: { backgroundColor?: string; color?: string } = {};
  if (bundle.bg) {
    style.backgroundColor = `rgb(${bundle.bg})`;
  }
  if (bundle.fg) {
    style.color = `rgb(${bundle.fg})`;
  }

  if (!linkify) {
    return React.createElement("span", { style, key }, bundle.content);
  }

  const words = bundle.content.split(" ").reduce(
    (words: React.ReactNode[], word: string, index: number) => {
      // If this isn't the first word, re-add the space removed from split.
      if (index !== 0) {
        words.push(" ");
      }

      // If  this isn't a link, just return the word as-is.
      if (!LINK_REGEX.test(word)) {
        words.push(word);
        return words;
      }

      words.push(
        React.createElement(
          "a",
          {
            key: index,
            href: word,
            target: "_blank"
          },
          `${word}`
        )
      );
      return words;
    },
    [] as React.ReactNode[]
  );
  return React.createElement("span", { style, key }, words);
}

declare type Props = {
  children: string;
  className?: string;
  linkify?: boolean;
};

export default function Ansi(props: Props) {
  return React.createElement(
    "code",
    { className: props.className },
    ansiToJSON(props.children).map(
      convertBundleIntoReact.bind(null, props.linkify)
    )
  );
}
