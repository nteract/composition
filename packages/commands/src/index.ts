import { COMMANDS } from "./commands";
import { makeCommandSetsComponent } from "./components";
import { CellTarget, CommandLocation, GapTarget, makeCellInfo, makeCellModelFromState, makeGapInfo, makeGapModel } from "./model";

export { COMMANDS };
export { CellTarget, CommandLocation, GapTarget, makeCellInfo, makeCellModelFromState, makeGapInfo, makeGapModel };

export const CellCommands =
  makeCommandSetsComponent(makeCellInfo, makeCellModelFromState, COMMANDS.CELL);

export const GapCommands =
  makeCommandSetsComponent(makeGapInfo, makeGapModel, COMMANDS.GAP);
