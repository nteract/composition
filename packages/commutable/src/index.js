/* @flow */

import type { Notebook as v4Notebook } from './v4';

import type {
  ImmutableNotebook,
  JSONType,
} from './types';

type PlaceholderNotebook = {
  nbformat: number,
  nbformat_minor: number,
};

const v4 = require('./v4');

const {
  emptyNotebook,
  emptyCodeCell,
  emptyMarkdownCell,
  appendCell,
  monocellNotebook,
} = require('./structures');

export type Notebook = PlaceholderNotebook & v4Notebook;

function freezeReviver(k: string, v: JSONType): JSONType {
  return Object.freeze(v);
}

// Expected usage of below is
// fromJS(parseNotebook(string|buffer))

function parseNotebook(notebookString: string): Notebook {
  return JSON.parse(notebookString, freezeReviver);
}

function fromJS(notebookJSON: Notebook): ImmutableNotebook {
  if (notebookJSON.nbformat === 4 && notebookJSON.nbformat_minor >= 0) {
    if (Array.isArray(notebookJSON.cells) && typeof notebookJSON.metadata === 'object') {
      return v4.fromJS(notebookJSON);
    }
  }

  if (notebookJSON.nbformat) {
    throw new TypeError(
      `nbformat v${notebookJSON.nbformat}.${notebookJSON.nbformat_minor} not recognized`
    );
  }

  throw new TypeError('This notebook format is not supported');
}

function toJS(immnb: ImmutableNotebook): v4Notebook {
  if (immnb.get('nbformat') === 4 && immnb.get('nbformat_minor') >= 0) {
    return v4.toJS(immnb);
  }
  throw new TypeError('Only notebook format 4 is supported');
}

// Expected usage is stringifyNotebook(toJS(immutableNotebook))
function stringifyNotebook(notebook: v4Notebook): string {
  return JSON.stringify(notebook, null, 2);
}

module.exports = {
  emptyCodeCell,
  emptyMarkdownCell,
  emptyNotebook,
  appendCell,
  monocellNotebook,
  toJS,
  fromJS,
  stringifyNotebook,
};

export type {
  ImmutableNotebook,
};
