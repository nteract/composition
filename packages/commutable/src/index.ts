import { Map as ImmutableMap } from "immutable";
import * as v4 from "./v4";
import * as v3 from "./v3";
import { ImmutableNotebook, JSONType } from "./types";

// .....................................
// API Exports
// Make sure the index.js.flow types stay in sync with this section
//

// from types
export * from "./types";

// from structures
export {
  emptyCodeCell,
  emptyMarkdownCell,
  emptyNotebook,
  monocellNotebook,
  createCodeCell,
  insertCellAt,
  insertCellAfter,
  deleteCell,
  appendCellToNotebook
} from "./structures";

// v4
export {
  StreamOutput,
  Output,
  createImmutableMimeBundle,
  createImmutableOutput
} from "./v4";

// general

export type Notebook = v4.Notebook | v3.Notebook;

const freezeReviver = <T extends JSONType>(k: string, v: T) =>
  Object.freeze(v) as T;

// Expected usage of below is fromJS(parseNotebook(string|buffer))
export const parseNotebook = (notebookString: string): Notebook =>
  JSON.parse(notebookString, freezeReviver);

export const fromJS = (
  notebook: Notebook | ImmutableNotebook
): ImmutableNotebook => {
  if (ImmutableMap.isMap(notebook)) {
    if (notebook.has("cellOrder") && notebook.has("cellMap")) {
      return notebook;
    }
    throw new TypeError(
      `commutable was passed an Immutable.Map structure that is not a notebook`
    );
  }

  if (notebook.nbformat === 4 && notebook.nbformat_minor >= 0) {
    if (
      Array.isArray(notebook.cells) &&
      typeof notebook.metadata === "object"
    ) {
      return v4.fromJS(notebook);
    }
  } else if (notebook.nbformat === 3 && notebook.nbformat_minor >= 0) {
    return v3.fromJS(notebook);
  }

  if (notebook.nbformat) {
    throw new TypeError(
      `nbformat v${notebook.nbformat}.${notebook.nbformat_minor} not recognized`
    );
  }

  throw new TypeError("This notebook format is not supported");
};

export const toJS = (immnb: ImmutableNotebook): v4.Notebook => {
  const minorVersion: null | number = immnb.get("nbformat_minor", null);

  if (
    immnb.get("nbformat") === 4 &&
    typeof minorVersion === "number" &&
    minorVersion >= 0
  ) {
    return v4.toJS(immnb);
  }
  throw new TypeError("Only notebook formats 3 and 4 are supported!");
};

// Expected usage is stringifyNotebook(toJS(immutableNotebook))
export const stringifyNotebook = (notebook: v4.Notebook) =>
  JSON.stringify(notebook, null, 2);
