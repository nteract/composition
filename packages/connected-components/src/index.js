// @flow

export {
  default as NewNotebookNavigation,
  PureNewNotebookNavigation
} from "./new-notebook-navigation";

export { default as NotebookMenu } from "./notebook-menu";
export { default as ModalController } from "./modal-controller";
