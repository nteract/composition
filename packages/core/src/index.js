//@flow
import * as actions from "./actions";
import * as actionTypes from "./actionTypes";
import * as middlewares from "./middlewares";
import * as reducers from "./reducers";
import * as epics from "./epics";
import * as selectors from "./selectors";
import * as state from "./state";

export * from "./state";

export { actions, actionTypes, middlewares, reducers, selectors, epics, state };
