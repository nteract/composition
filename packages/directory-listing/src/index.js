// @flow
import { Entry } from "./components/entry";
import { Listing } from "./components/listing";
import { Icon } from "./components/icon";
import { Name } from "./components/name";
import { LastSaved } from "./components/lastsaved";

export { Entry, Listing, Icon, Name, LastSaved };
