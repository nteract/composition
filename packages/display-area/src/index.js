// @flow
import Display from "./display";
import RichestMime from "./richest-mime";

module.exports = {
  Display,
  RichestMime
};
