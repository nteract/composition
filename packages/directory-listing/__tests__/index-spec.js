// @flow
import * as React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

import { Entry, Listing, Name, Icon, LastSaved } from "../src";

describe("Listing", () => {
  it("accepts props and renders entries in directory in a table", () => {
    const component = (
      <Listing>
        <Entry>
          <Icon fileType={"directory"} />
          <Name>{"linky"}</Name>
          <LastSaved
            last_modified={
              "Fri Jun 22 2018 00:15:55 GMT-0400 (Eastern Daylight Time)"
            }
          />
        </Entry>
      </Listing>
    );
    expect(toJson(component)).toMatchSnapshot();
  });
});

describe("Entry", () => {
  it("accepts props and renders entries in directory", () => {
    const component = shallow(
      <Entry>
        <Icon fileType={"directory"} />
        <Name>{"linky jr"}</Name>
        <LastSaved
          last_modified={
            "Fri Jun 22 2018 00:15:55 GMT-0400 (Eastern Daylight Time)"
          }
        />
      </Entry>
    );
    expect(toJson(component)).toMatchSnapshot();
  });
});
