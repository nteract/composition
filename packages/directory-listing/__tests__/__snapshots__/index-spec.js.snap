// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entry accepts props and renders entries in directory 1`] = `
<tr
  className="jsx-3511019751 directory-entry"
>
  <Icon
    className="directory-entry-field"
    color="#0366d6"
    fileType="directory"
    key=".0"
  />
  <Name
    className="directory-entry-field"
    key=".1"
  >
    linky jr
  </Name>
  <LastSaved
    className="directory-entry-field"
    key=".2"
    lastModified={null}
    last_modified="Fri Jun 22 2018 00:15:55 GMT-0400 (Eastern Daylight Time)"
  />
  <JSXStyle
    css="tr.jsx-3511019751{border-top:1px solid #eaecef;}tr.jsx-3511019751:hover{background-color:#f6f8fa;-webkit-transition:background-color 0.1s ease-out;transition:background-color 0.1s ease-out;}tr.jsx-3511019751:first-child{border-top:none;}tr.jsx-3511019751:last-child{border-bottom:none;}"
    styleId="3511019751"
  />
</tr>
`;

exports[`Listing accepts props and renders entries in directory in a table 1`] = `undefined`;
